text/0126-introduce-xcq.md

```diff
@@ -0,0 +1,354 @@
+
+```rust
+linker
+    .define_typed(
+        "host_call",
+        move |caller: Caller<'_, Self::UserData>,
+              extension_id: u64,
+              call_ptr: u32,
+              call_len: u32|
+              -> Result<u64, ExtensionError> {
+                  ...
+              });
+```
+
+- `PermissionController`: Filters guest XCQ program calling requests, useful for host chains to disable some queries by filtering invoking sources.
+
+```rust
+pub trait PermissionController {
+    fn is_allowed(extension_id: ExtensionIdTy, call: &[u8], source: InvokeSource) -> bool;
+}
+#[derive(Copy, Clone)]
+pub enum InvokeSource {
+    RuntimeAPI,
+    XCM,
+    Extrinsic,
+    Runtime,
+}
+```
+
+### XCQ Program Structure
+
+An XCQ program is structured as a PolkaVM program with the following key components:
+
+- Imported Functions:
+  - `host_call`: Dispatches call requests to the XCQ Extension Executor.
+
+     ```rust
+     #[polkavm_derive::polkavm_import]
+     extern "C" {
+         fn host_call(extension_id: u64, call_ptr: u32, call_len: u32) -> u64;
+     }
+     ```
+
+     Results are SCALE-encoded bytes, with the pointer address (lower 32 bits) and length (higher 32 bits) packed into a u64.
+
+  - `return_ty`: Returns the type of the function call result.
+
+     ```rust
+     #[polkavm_derive::polkavm_import]
+     extern "C" {
+         fn return_ty(extension_id: u64, call_index: u32) -> u64;
+     }
+     ```
+
+     Results are SCALE-encoded bytes, with the pointer address and length packed similarly to `host_call`.
+
+- Exported Functions:
+  - `main`: The entry point of the XCQ program. It performs type checking, arguments and results passing and executes the query.
+
+### XCQ Program Execution Flow
+
+The interaction between an XCQ program and the XCQ Extension Executor follows these steps:
+
+1. Program Loading: The Executor loads the PolkaVM program binary.
+
+2. Environment Setup: The Executor configures the PolkaVM environment, registering host functions like `host_call` and `return_ty`.
+
+3. Main Function Execution: The Executor calls the program's `main` function, passing serialized query arguments.
+
+4. Program Execution:
+   1. Type Checking: The program uses the `return_ty` function to ensure compatibility with supported chain extensions.
+   2. Query Execution: The program executes the query using `host_call` and performs custom computations.
+   3. Result Serialization: The program serializes the result, writes it to shared memory, and returns the pointer and length to the executor.
+
+5. Result Retrieval: The Executor reads the result from shared memory and returns it to the caller.
+
+### XCM integration
+
+The integration of XCQ into XCM is acheived by adding a new instruction to XCM, as well as a new variant of the `Response` type in `QueryResponse` message.:
+
+- A new `ReportQuery` instruction
+
+```rust
+ReportQuery {
+  query: SizeLimitedXcq,
+  weight_limit: Option<Weight>,
```

Comment:
How can someone get this weight? Is it easy? If not, we could just limit it by the overall fee paid in `PayFees`?
We need to have a good way of using this, otherwise everyone will just use `Unlimited`

Reply:
maybe we should modify `execute_query` to have it also return used weight so it can be used to estimate weight usage
We could make `Unlimited` weight means the weight to be limited by fees. Limited weight will be useful for the case that we need to reserve some fees for other instruction, such as transact